microsetta_interface/templates/source.jinja2

```diff
@@ -9,320 +16,143 @@
-                    validateOptions["submitHandler"] = submitHandler;
-                }
-                $(form_selector).validate(validateOptions);
-            };
-        }
+    <script type="text/javascript">
+        function takeExternalSurvey(survey_link, survey_title, survey_description) {
+            $("#modal-title").empty();
+            $("#modal-title").text(survey_title);
+            $("#survey-description").empty();
+            $("#survey-description").text(survey_description);
+            $("#take-survey").click(async function(e) {
+                window.open(survey_link, '_blank').focus();
+                $("#source_modal").modal('hide');
 
-        let kitSubmitHandler = function(form, evt){
-            let samples_selector = '#samples';
-            let error_txt = "";
-            let successful = false;
-            $(form).ajaxSubmit({
-                success: function (samples, textStatus, jqXHR) {
-                    $(samples_selector).empty();
-                    for (let i = 0; i < samples.length; i++) {
-                        let bc = samples[i].sample_barcode;
-                        let sample_id = samples[i].sample_id;
-                        let item = $('<div class="container list-group-item">' +
-                                     '<div class="row">' +
-                                     '<div class="col-sm">' +
-                                     '<label for="' + bc + '">' + bc + '</label>' +
-                                     '</div>' +
-                                     '<div class="col-sm">' +
-                                     '<input type="checkbox" name="sample_id" id="' + bc + '" value="' + sample_id + '"/>' +
-                                     '</div>' +
-                                     '</div>');
-                        $(samples_selector).append(item);
-                    }
-                    let div = $('#choose_sample_div');
-                    div.collapse('show');
-                    successful = true;
-                },
-                error: function (jqXHR, textStatus, errorThrown) {
-                    error_txt = jqXHR.responseText;
-                },
-                complete: function (data, textStatus, outputObj) {
-                    if (successful)
-                        return;
-                    if (error_txt === "")
-                        error_txt = textStatus;
-                    // Get customized error page that will show whatever error message we captured
-                    // here in the front end
-                    let queryParam = encodeURIComponent(error_txt);
-                    window.location.replace("{{ endpoint }}/error?error_msg=" + queryParam);
-                }
+                // We need to wait a couple of seconds for the external survey to register in the database, then reload the page. This is not ideal.
+                await new Promise(r => setTimeout(r, 2000));
```

Comment:
I don't think there was an associated code change here -- do you know how bad this would be if the 2s window is hit?

Reply:
At the moment, it's fairly minor - the only consequence is that the person would be able to re-take an external survey with a different unique ID. And that case would only occur if they close the newly-opened tab with an external survey, reopen the external survey's modal, and then click the button to take the link before the page reloads. If we ever have external surveys with a cost attached, it could be problematic, but in current usage, it's a minor annoyance.